### PR TITLE
feat: ompi images for use with slurm

### PR DESCRIPTION
**Pull request by stoksc**
## Description
Subject: Add OpenMPI to CPU images on slurm-images branch (https://github.com/determined-ai/environments/pull/142)
These mods are essentially the same as those in the PR for branch
feature/cpu-mpi, which originally targeted master, but this mod
targets the slurm-images branch.  This mod also used the newer
Makefile, which resolves several conflicts that existed with the
original feature/cpu-mpi branch.  Each of the changes are described
below using the log from the original commits.

Subject: Add OpenMPI support to NVIDIA GPU targets (https://github.com/determined-ai/environments/pull/144)
Description:

    This mod adds OpenMPI support to the targets for NVIDIA gpus, similar
    to what was done for the CPU images. OpenMPI can now be built into the
    base image using the USE_MPI=1 option, such as:

    $> make USE_MPI="1" build-tf24-gpu

    After building this image and converting to Singularity, the image was
    tested on an IB cluster using the OSU benchmarks to verify
    performance.

    $ srun --mpi=pmi2 -p gpuNodes -n 2 --ntasks-per-node=1 singularity
    exec --bind /lus /lus/scratch/username/detAI-cuda-11.1-torch-1.9-tf-2.4-gpu-mpi.sif ./osu-micro-benchmarks-5.6.2/mpi/one-sided/osu_get_bw
    1                       2.06
    2                       4.21
    4                       8.41
    8                      16.83
    16                     33.77
    32                     66.42
    64                    128.51
    128                   246.12
    256                   493.93
    512                  1013.98
    1024                 1871.85
    2048                 3200.73
    4096                 4763.36
    8192                 6239.26
    16384                8293.17
    32768               11066.21
    65536               11517.27
    131072              11574.10
    262144              11581.30
    524288              11165.66
    1048576             10911.13
    2097152             10533.11
    4194304              9755.71

    Also tested with the TF2 MNIST test on 2 nodes using 8 gpus per node:

    $ srun --mpi=pmi2 -n 16 --ntasks-per-node=8 singularity exec --nv
    --bind /lus /lus/scratch/username/detAI-cuda-11.1-torch-1.9-tf-2.4-gpu-mpi.sif
    python /lus/scratch/username/tensorflow2_mnist.py >& log-gpu.txt

    In the log-gpu.txt file I verified that hvd.init() showed the correct
    MPI world size and ranks:

    $ grep "hvd\.size" log-gpu.txt
    hvd.local_rank(): 6, hvd.rank(): 6, hvd.size(): 16
    hvd.local_rank(): 2, hvd.rank(): 2, hvd.size(): 16
    hvd.local_rank(): 4, hvd.rank(): 12, hvd.size(): 16
    hvd.local_rank(): 6, hvd.rank(): 14, hvd.size(): 16
    hvd.local_rank(): 2, hvd.rank(): 10, hvd.size(): 16
    hvd.local_rank(): 4, hvd.rank(): 4, hvd.size(): 16
    hvd.local_rank(): 3, hvd.rank(): 11, hvd.size(): 16
    hvd.local_rank(): 0, hvd.rank(): 0, hvd.size(): 16
    hvd.local_rank(): 7, hvd.rank(): 7, hvd.size(): 16
    hvd.local_rank(): 0, hvd.rank(): 8, hvd.size(): 16
    hvd.local_rank(): 7, hvd.rank(): 15, hvd.size(): 16
    hvd.local_rank(): 3, hvd.rank(): 3, hvd.size(): 16
    hvd.local_rank(): 1, hvd.rank(): 9, hvd.size(): 16
    hvd.local_rank(): 5, hvd.rank(): 5, hvd.size(): 16
    hvd.local_rank(): 5, hvd.rank(): 13, hvd.size(): 16
    hvd.local_rank(): 1, hvd.rank(): 1, hvd.size(): 16

    The tensorflow2_mnist.py test completed as expected:

    $ tail log-gpu.txt
    Step #580          Loss: 0.049359
    Step #580          Loss: 0.064985
    Step #590          Loss: 0.040959
    Step #590          Loss: 0.183963
    Step #600          Loss: 0.039476
    Step #600          Loss: 0.103435
    Step #610          Loss: 0.049162
    Step #610          Loss: 0.048544
    Step #620          Loss: 0.084336
    Step #620          Loss: 0.046733

    Note: I was not able to build the tf1-gpu and tf2-gpu images because I
    encountered an error in the install.apex.sh similar to:

    ERROR: Command errored out with exit status 1: /opt/conda/bin/python
    -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] =
    '"'"'/tmp/pip-req-build-ao9mgp98/setup.py'"'"';
    __file__='"'"'/tmp/pip-req-build-ao9mgp98/setup.py'"'"';f =
    getattr(tokenize, '"'"'open'"'"', open)(__file__) if
    os.path.exists(__file__) else io.StringIO('"'"'from setuptools import
    setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"',
    '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))'
    --cpp_ext --cuda_ext install --record
    /tmp/pip-record-4l_4so8t/install-record.txt
    --single-version-externally-managed --compile --install-headers
    /opt/conda/include/python3.8/apex Check the logs for full command
    output.
    The command '/bin/sh -c /tmp/det_dockerfile_scripts/install_apex.sh'
    returned a non-zero code: 1
    Makefile:229: recipe for target 'build-tf2-gpu' failed

    This error also occurs when I try building on the current slurm-images
    branch without any of the changes from this PR.

ci: add jobs to build and publish MPI images

## Checklist

- [ ] Bump VERSION to make the pushed images are tagged with the right version.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -52,8 +74,9 @@ export AWS_MAX_ATTEMPTS=360
 .PHONY: build-cpu-py-37-base build-cpu-py-38-base  build-gpu-cuda-111-base build-gpu-cuda-112-base build-gpu-cuda-113-base
 build-cpu-py-37-base:
 	docker build -f Dockerfile-base-cpu \
-		--build-arg BASE_IMAGE="ubuntu:18.04" \
+		--build-arg BASE_IMAGE="$(UBUNTU_IMAGE_TAG)" \
```
Comment (stoksc) [inline Makefile:77]: this is the only change to existing images -- we need to bump to ubuntu:20.04 for these images

File: Dockerfile-base-cpu [hunk 2]
```diff
@@ -19,13 +19,17 @@ RUN apt-get update \
 		libtool \
 		git \
 		krb5-user \
-		g++-4.8 \
+		g++ \
+		cmake \
```
Comment (mpkouznetsov) [inline Dockerfile-base-cpu:23]: Do we need this, given that we `pip install cmake==3.22.4` in Dockerfile-default-xxx?

Comment (mpkouznetsov) [reply Dockerfile-base-cpu:23]: oh, I guess we do if ompi.sh needs it and is run as part of building the base images.

**Review by mpkouznetsov — APPROVED** [review]
LGTM. It would be good to let the build run and check the images generated by the latest commit. I am happy to verify the arm64 image if it is built.

**Comment by mpkouznetsov** [comment]
I confirmed arm64 is working.

**Comment by stoksc** [comment]
bumped to coda images from 11.1 to 11.1.1 due to the images missing (?). maybe due to https://nvd.nist.gov/vuln/detail/CVE-2020-5991, but idk, could not dig up an explanation.

**Review by mpkouznetsov — APPROVED** [review]

**Review by cdrickett — APPROVED** [review]
Looks good. Thanks for finishing this PR.

**Pull request merged**